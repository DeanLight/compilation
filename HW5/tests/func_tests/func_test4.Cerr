[[[ added func: print with 1 params and label: ]]]
[[[ added func: printi with 1 params and label: ]]]
print param:2(1)
printi param:3(1)
[FIRST_PROGRAM_POINT]
[set_func_label] label print for function print
[set_func_label] label printi for function printi
(END_OF)  [FIRST_PROGRAM_POINT]
Lex Ate token void__
Lex Ate token  id__f__
Lex Ate token  (__
Lex Ate token bool__
Lex Ate token  id__cond__
<<FormalDecl_Semantic: id[cond]>>
<<adding param: [cond] of type [5]>>Lex Ate token )__
<<FormalsSem>><<FuncHead_semantic, id:[f]>>
[[[ added func: f with 1 params and label: ]]]
FuncHead_IR for func: f
got label label_29 in line: -1
[set_func_label] label label_29 for function f
[get_func_label] f's label is: label_29
written label: label_29
END_OF [FuncHead_IR] 
Lex Ate token  {__
Lex Ate token if__
Lex Ate token  (__
Lex Ate token  id__cond__
Lex Ate token )__
<<[Exp_Semantic_Id] id:[cond]>>
[Exp_IR] Exp -> id: cond
END_OF [Exp_IR] Exp -> id
[SJ_Exp_IR]: cond
[MARKER IR]: label: bp_label_0
Lex Ate token  {__
Lex Ate token  id__print__
Lex Ate token  (__
CallHeader_IR
[Emitter] store_registers: 0
END_OF CallHeader_IR
Lex Ate token  str__"boy\n"__
[Exp_IR] Exp -> string"boy\n"
END_OF [Exp_IR] Exp -> string
Lex Ate token )__
<<ExpList_Semantic>> explist -> exp : 
exp is of type STRING
[ExpList_IR: ExpList->Exp]: 
<<CallSemantics [ID(Exp)] >>
tmpParams: 1
tmpActP: 1
tmptmp: 1
[Call_IR] : with arguments print
[get_func_label] print's label is: print
found a string typed parameter print
[Emitter] REstore_registers: 0
END_OF [Call_IR]
Lex Ate token  ;__
[Statement_IR Call] nestlistSize: 0
[Statement_IR Call] nestlistSize: 0
[Statement_IR: States->SignleState]  nextlistSize=0
Lex Ate token  id__f__
[MARKER IR]: label: bp_label_1
Lex Ate token  (__
CallHeader_IR
[Emitter] store_registers: 0
END_OF CallHeader_IR
Lex Ate token false__
Lex Ate token )__
<<ExpList_Semantic>> explist -> exp : false
exp is of type BOOL
[ExpList_IR: ExpList->Exp]: false
<<CallSemantics [ID(Exp)] >>
tmpParams: 1
tmpActP: 1
tmptmp: 1
[Call_IR] : with arguments f
[get_func_label] f's label is: label_29
[Emitter] REstore_registers: 0
END_OF [Call_IR]
Lex Ate token  ;__
[Statement_IR Call] nestlistSize: 0
[Statement_IR Call] nestlistSize: 0
[Statement_IR: States -> states1 M singleState]
Num of nextlist to bp: 0 with label: bp_label_1 breaklist size is 0
Lex Ate token  }__
[MARKER IR]: label: bp_label_2
Lex Ate token else__
Lex Ate token  {__
Lex Ate token  id__print__
Lex Ate token  (__
CallHeader_IR
[Emitter] store_registers: 0
END_OF CallHeader_IR
Lex Ate token  str__"girl\n"__
[Exp_IR] Exp -> string"girl\n"
END_OF [Exp_IR] Exp -> string
Lex Ate token )__
<<ExpList_Semantic>> explist -> exp : 
exp is of type STRING
[ExpList_IR: ExpList->Exp]: 
<<CallSemantics [ID(Exp)] >>
tmpParams: 1
tmpActP: 1
tmptmp: 1
[Call_IR] : with arguments print
[get_func_label] print's label is: print
found a string typed parameter print
[Emitter] REstore_registers: 0
END_OF [Call_IR]
Lex Ate token  ;__
[Statement_IR Call] nestlistSize: 0
[Statement_IR Call] nestlistSize: 0
[Statement_IR: States->SignleState]  nextlistSize=0
Lex Ate token  }__
size of next list for if is0
size of next list for else is0
[Statement_next_patcher_IR] previously held nextlist size:0
[Statement_next_patcher_IR] currently held nextlist size:1
[Statement_IR: States->SignleState]  nextlistSize=1
Lex Ate token  }__
[FuncDeclNode_IR]: 
checking retType of func: f
If void, an extra ret will be added
Ret type: 1
Lex Ate token bool__
Lex Ate token  id__g__
Lex Ate token  (__
Lex Ate token bool__
Lex Ate token  id__gen__
<<FormalDecl_Semantic: id[gen]>>
<<adding param: [gen] of type [5]>>Lex Ate token )__
<<FormalsSem>><<FuncHead_semantic, id:[g]>>
[[[ added func: g with 1 params and label: ]]]
FuncHead_IR for func: g
got label label_135 in line: -1
[set_func_label] label label_135 for function g
[get_func_label] g's label is: label_135
written label: label_135
END_OF [FuncHead_IR] 
Lex Ate token  {__
Lex Ate token if__
Lex Ate token  (__
Lex Ate token  id__gen__
Lex Ate token )__
<<[Exp_Semantic_Id] id:[gen]>>
[Exp_IR] Exp -> id: gen
END_OF [Exp_IR] Exp -> id
[SJ_Exp_IR]: gen
[MARKER IR]: label: bp_label_5
Lex Ate token  {__
Lex Ate token  id__print__
Lex Ate token  (__
CallHeader_IR
[Emitter] store_registers: 0
END_OF CallHeader_IR
Lex Ate token  str__"M "__
[Exp_IR] Exp -> string"M "
END_OF [Exp_IR] Exp -> string
Lex Ate token )__
<<ExpList_Semantic>> explist -> exp : 
exp is of type STRING
[ExpList_IR: ExpList->Exp]: 
<<CallSemantics [ID(Exp)] >>
tmpParams: 1
tmpActP: 1
tmptmp: 1
[Call_IR] : with arguments print
[get_func_label] print's label is: print
found a string typed parameter print
[Emitter] REstore_registers: 0
END_OF [Call_IR]
Lex Ate token  ;__
[Statement_IR Call] nestlistSize: 0
[Statement_IR Call] nestlistSize: 0
[Statement_IR: States->SignleState]  nextlistSize=0
Lex Ate token return__
[MARKER IR]: label: bp_label_6
Lex Ate token true__
Lex Ate token  ;__
<<RetExp, ExpT:[5] scopeT:[5]>>
[Statement_IR] Return noneVoid
END_OF [Statement_IR] Return noneVoid
[Statement_IR: States -> states1 M singleState]
Num of nextlist to bp: 0 with label: bp_label_6 breaklist size is 0
Lex Ate token  }__
[MARKER IR]: label: bp_label_7
Lex Ate token else__
Lex Ate token  {__
Lex Ate token  id__print__
Lex Ate token  (__
CallHeader_IR
[Emitter] store_registers: 0
END_OF CallHeader_IR
Lex Ate token  str__"F "__
[Exp_IR] Exp -> string"F "
END_OF [Exp_IR] Exp -> string
Lex Ate token )__
<<ExpList_Semantic>> explist -> exp : 
exp is of type STRING
[ExpList_IR: ExpList->Exp]: 
<<CallSemantics [ID(Exp)] >>
tmpParams: 1
tmpActP: 1
tmptmp: 1
[Call_IR] : with arguments print
[get_func_label] print's label is: print
found a string typed parameter print
[Emitter] REstore_registers: 0
END_OF [Call_IR]
Lex Ate token  ;__
[Statement_IR Call] nestlistSize: 0
[Statement_IR Call] nestlistSize: 0
[Statement_IR: States->SignleState]  nextlistSize=0
Lex Ate token return__
[MARKER IR]: label: bp_label_8
Lex Ate token false__
Lex Ate token  ;__
<<RetExp, ExpT:[5] scopeT:[5]>>
[Statement_IR] Return noneVoid
END_OF [Statement_IR] Return noneVoid
[Statement_IR: States -> states1 M singleState]
Num of nextlist to bp: 0 with label: bp_label_8 breaklist size is 0
Lex Ate token  }__
size of next list for if is0
size of next list for else is0
[Statement_next_patcher_IR] previously held nextlist size:0
[Statement_next_patcher_IR] currently held nextlist size:1
[Statement_IR: States->SignleState]  nextlistSize=1
Lex Ate token  }__
[FuncDeclNode_IR]: 
checking retType of func: g
If void, an extra ret will be added
Ret type: 5
Lex Ate token void__
Lex Ate token  id__main__
Lex Ate token  (__
Lex Ate token )__
<<FuncHead_semantic, id:[main]>>
[[[ added func: main with 0 params and label: ]]]
FuncHead_IR for func: main
got label label_224 in line: -1
[set_func_label] label label_224 for function main
Saving main label: label_224
[get_func_label] main's label is: label_224
written label: label_224
END_OF [FuncHead_IR] 
Lex Ate token  {__
Lex Ate token  id__print__
Lex Ate token  (__
CallHeader_IR
[Emitter] store_registers: 0
END_OF CallHeader_IR
Lex Ate token  str__"apple \n"__
[Exp_IR] Exp -> string"apple \n"
END_OF [Exp_IR] Exp -> string
Lex Ate token )__
<<ExpList_Semantic>> explist -> exp : 
exp is of type STRING
[ExpList_IR: ExpList->Exp]: 
<<CallSemantics [ID(Exp)] >>
tmpParams: 1
tmpActP: 1
tmptmp: 1
[Call_IR] : with arguments print
[get_func_label] print's label is: print
found a string typed parameter print
[Emitter] REstore_registers: 0
END_OF [Call_IR]
Lex Ate token  ;__
[Statement_IR Call] nestlistSize: 0
[Statement_IR Call] nestlistSize: 0
[Statement_IR: States->SignleState]  nextlistSize=0
Lex Ate token  id__f__
[MARKER IR]: label: bp_label_11
Lex Ate token  (__
CallHeader_IR
[Emitter] store_registers: 0
END_OF CallHeader_IR
Lex Ate token  id__g__
Lex Ate token  (__
CallHeader_IR
[Emitter] store_registers: 0
END_OF CallHeader_IR
Lex Ate token true__
Lex Ate token )__
<<ExpList_Semantic>> explist -> exp : true
exp is of type BOOL
[ExpList_IR: ExpList->Exp]: true
<<CallSemantics [ID(Exp)] >>
tmpParams: 1
tmpActP: 1
tmptmp: 1
[Call_IR] : with arguments g
[get_func_label] g's label is: label_135
[Emitter] REstore_registers: 0
END_OF [Call_IR]
[Exp_IR] Exp->callg
[Exp_IR] endof Exp->callg
Lex Ate token )__
<<ExpList_Semantic>> explist -> exp : g
exp is of type BOOL
[ExpList_IR: ExpList->Exp]: g
<<CallSemantics [ID(Exp)] >>
tmpParams: 1
tmpActP: 1
tmptmp: 1
[Call_IR] : with arguments f
[get_func_label] f's label is: label_29
[Emitter] REstore_registers: 0
END_OF [Call_IR]
Lex Ate token  ;__
[Statement_IR Call] nestlistSize: 0
[Statement_IR Call] nestlistSize: 0
[Statement_IR: States -> states1 M singleState]
Num of nextlist to bp: 0 with label: bp_label_11 breaklist size is 0
Lex Ate token  id__f__
[MARKER IR]: label: bp_label_15
Lex Ate token  (__
CallHeader_IR
[Emitter] store_registers: 0
END_OF CallHeader_IR
Lex Ate token  id__g__
Lex Ate token  (__
CallHeader_IR
[Emitter] store_registers: 0
END_OF CallHeader_IR
Lex Ate token false__
Lex Ate token )__
<<ExpList_Semantic>> explist -> exp : false
exp is of type BOOL
[ExpList_IR: ExpList->Exp]: false
<<CallSemantics [ID(Exp)] >>
tmpParams: 1
tmpActP: 1
tmptmp: 1
[Call_IR] : with arguments g
[get_func_label] g's label is: label_135
[Emitter] REstore_registers: 0
END_OF [Call_IR]
[Exp_IR] Exp->callg
[Exp_IR] endof Exp->callg
Lex Ate token )__
<<ExpList_Semantic>> explist -> exp : g
exp is of type BOOL
[ExpList_IR: ExpList->Exp]: g
<<CallSemantics [ID(Exp)] >>
tmpParams: 1
tmpActP: 1
tmptmp: 1
[Call_IR] : with arguments f
[get_func_label] f's label is: label_29
[Emitter] REstore_registers: 0
END_OF [Call_IR]
Lex Ate token  ;__
[Statement_IR Call] nestlistSize: 0
[Statement_IR Call] nestlistSize: 0
[Statement_IR: States -> states1 M singleState]
Num of nextlist to bp: 0 with label: bp_label_15 breaklist size is 0
Lex Ate token  }__
[FuncDeclNode_IR]: 
checking retType of func: main
If void, an extra ret will be added
Ret type: 1
[Program_IR] backpatching line number 4 with [get_func_label] main's label is: label_224
label_224
[get_func_label] main's label is: label_224
ignore me1 (label_224)
[END_OF Program_IR]
