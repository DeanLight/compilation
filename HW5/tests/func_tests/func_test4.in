

void f(bool cond){
	if(cond){
		print("boy\n");
		f(false);
	}else{
		print("girl\n");
	}
	
}	

bool g(bool gen){
	if(gen){
		print("M "); 
		return true;
	}
	else{
		print("F ");
		return false;
	}
}


void main(){
	print("apple \n");
	f(g(true));
	f(g(false));
}