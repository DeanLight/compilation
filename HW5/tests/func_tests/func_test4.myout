Loaded: ./exceptions.s
apple 
M boy
girl
girl
F girl
