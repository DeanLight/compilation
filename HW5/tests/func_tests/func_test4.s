.data
pvufne1:	.asciiz "Error division by zero
"
string_label_0:	.asciiz "boy\n"
string_label_1:	.asciiz "boy\n"
string_label_2:	.asciiz "girl\n"
string_label_3:	.asciiz "girl\n"
string_label_4:	.asciiz "M "
string_label_5:	.asciiz "M "
string_label_6:	.asciiz "F "
string_label_7:	.asciiz "F "
string_label_8:	.asciiz "apple \n"
string_label_9:	.asciiz "apple \n"
.text
	#first program point
main:
	#initialize fp
	move	$fp,$sp
	jal	label_224
	#exiting...
li $v0, 10
syscall
	#print_func:
print:
lw $a0,4($sp)
li $v0,4
syscall
jr $ra
	#printi func:
printi:
lw $a0,4($sp)
li $v0,1
syscall
jr $ra
	#div by 0 handler:
nayalo:
	li	$v0, 4
	la	$a0, pvufne1
	syscall
	#exiting...
li $v0, 10
syscall
	#Func f:
label_29:
	#Getting Var falue for [Exp->id]: cond offset is 4($fp)
	#			 __allocating reg $t0
	lw $t0, 4($fp)
	#a Bool Var cond in boolean operator
	#If true
	bne $t0, $zero,bp_label_0
	#			 __freeing reg $t0
	#If False
	j	bp_label_2
	#marker Label
bp_label_0:
	#func header store regs before call
	#stored 0 registers
	#caught a string
	#preparing to call print
	sw $fp, ($sp)
	addiu $sp, $sp, -4
	sw $ra, ($sp)
	addiu $sp, $sp, -4
	#loading str address to stack
	la $v0, string_label_1
	sw $v0, ($sp)
	addiu $sp, $sp, -4
	move	$fp,$sp
	#jumping to func: print
	jal	print
	#returned from func: print
	#poping 1 params from stack 
	addiu $sp, $sp, 4
	#restoring ra and fp
	addiu $sp, $sp, 4
	lw $ra, ($sp)
	addiu $sp, $sp, 4
	lw $fp, ($sp)
	#restoring 0 previously used registers
	#Moving funcRes (if exists) to next free register
	#finished calling print
	#marker Label
bp_label_1:
	#func header store regs before call
	#stored 0 registers
	#exp derived false
	#Moving False into new Param Reg
	#			 __allocating reg $t0
	li	$t0,0
	#preparing to call f
	sw $fp, ($sp)
	addiu $sp, $sp, -4
	sw $ra, ($sp)
	addiu $sp, $sp, -4
	#pushing reg $t0 to stack
	sw $t0, ($sp)
	addiu $sp, $sp, -4
	#			 __freeing reg $t0
	move	$fp,$sp
	#jumping to func: f
	jal	label_29
	#returned from func: f
	#poping 1 params from stack 
	addiu $sp, $sp, 4
	#restoring ra and fp
	addiu $sp, $sp, 4
	lw $ra, ($sp)
	addiu $sp, $sp, 4
	lw $fp, ($sp)
	#restoring 0 previously used registers
	#Moving funcRes (if exists) to next free register
	#finished calling f
	#marker Label
bp_label_2:
	#func header store regs before call
	#stored 0 registers
	#caught a string
	#preparing to call print
	sw $fp, ($sp)
	addiu $sp, $sp, -4
	sw $ra, ($sp)
	addiu $sp, $sp, -4
	#loading str address to stack
	la $v0, string_label_3
	sw $v0, ($sp)
	addiu $sp, $sp, -4
	move	$fp,$sp
	#jumping to func: print
	jal	print
	#returned from func: print
	#poping 1 params from stack 
	addiu $sp, $sp, 4
	#restoring ra and fp
	addiu $sp, $sp, 4
	lw $ra, ($sp)
	addiu $sp, $sp, 4
	lw $fp, ($sp)
	#restoring 0 previously used registers
	#Moving funcRes (if exists) to next free register
	#finished calling print
	#end of ifelse
bp_label_3:
	#end of statement jump
	j	bp_label_4
bp_label_4:
	#Adding an extre return just in case
	move	$sp,$fp
	jr	$ra
	#Func g:
label_135:
	#Getting Var falue for [Exp->id]: gen offset is 4($fp)
	#			 __allocating reg $t0
	lw $t0, 4($fp)
	#a Bool Var gen in boolean operator
	#If true
	bne $t0, $zero,bp_label_5
	#			 __freeing reg $t0
	#If False
	j	bp_label_7
	#marker Label
bp_label_5:
	#func header store regs before call
	#stored 0 registers
	#caught a string
	#preparing to call print
	sw $fp, ($sp)
	addiu $sp, $sp, -4
	sw $ra, ($sp)
	addiu $sp, $sp, -4
	#loading str address to stack
	la $v0, string_label_5
	sw $v0, ($sp)
	addiu $sp, $sp, -4
	move	$fp,$sp
	#jumping to func: print
	jal	print
	#returned from func: print
	#poping 1 params from stack 
	addiu $sp, $sp, 4
	#restoring ra and fp
	addiu $sp, $sp, 4
	lw $ra, ($sp)
	addiu $sp, $sp, 4
	lw $fp, ($sp)
	#restoring 0 previously used registers
	#Moving funcRes (if exists) to next free register
	#finished calling print
	#marker Label
bp_label_6:
	#exp derived true
	#return noneVoid in v0
	li	$v0,1
	move	$sp,$fp
	jr	$ra
	#marker Label
bp_label_7:
	#func header store regs before call
	#stored 0 registers
	#caught a string
	#preparing to call print
	sw $fp, ($sp)
	addiu $sp, $sp, -4
	sw $ra, ($sp)
	addiu $sp, $sp, -4
	#loading str address to stack
	la $v0, string_label_7
	sw $v0, ($sp)
	addiu $sp, $sp, -4
	move	$fp,$sp
	#jumping to func: print
	jal	print
	#returned from func: print
	#poping 1 params from stack 
	addiu $sp, $sp, 4
	#restoring ra and fp
	addiu $sp, $sp, 4
	lw $ra, ($sp)
	addiu $sp, $sp, 4
	lw $fp, ($sp)
	#restoring 0 previously used registers
	#Moving funcRes (if exists) to next free register
	#finished calling print
	#marker Label
bp_label_8:
	#exp derived false
	#return noneVoid in v0
	li	$v0,0
	move	$sp,$fp
	jr	$ra
	#end of ifelse
bp_label_9:
	#end of statement jump
	j	bp_label_10
bp_label_10:
	#Adding an extre return just in case
	move	$sp,$fp
	jr	$ra
	#Func main:
label_224:
	#func header store regs before call
	#stored 0 registers
	#caught a string
	#preparing to call print
	sw $fp, ($sp)
	addiu $sp, $sp, -4
	sw $ra, ($sp)
	addiu $sp, $sp, -4
	#loading str address to stack
	la $v0, string_label_9
	sw $v0, ($sp)
	addiu $sp, $sp, -4
	move	$fp,$sp
	#jumping to func: print
	jal	print
	#returned from func: print
	#poping 1 params from stack 
	addiu $sp, $sp, 4
	#restoring ra and fp
	addiu $sp, $sp, 4
	lw $ra, ($sp)
	addiu $sp, $sp, 4
	lw $fp, ($sp)
	#restoring 0 previously used registers
	#Moving funcRes (if exists) to next free register
	#finished calling print
	#marker Label
bp_label_11:
	#func header store regs before call
	#stored 0 registers
	#func header store regs before call
	#stored 0 registers
	#exp derived true
	#Moving True into new Param Reg
	#			 __allocating reg $t0
	li	$t0,1
	#preparing to call g
	sw $fp, ($sp)
	addiu $sp, $sp, -4
	sw $ra, ($sp)
	addiu $sp, $sp, -4
	#pushing reg $t0 to stack
	sw $t0, ($sp)
	addiu $sp, $sp, -4
	#			 __freeing reg $t0
	move	$fp,$sp
	#jumping to func: g
	jal	label_135
	#returned from func: g
	#poping 1 params from stack 
	addiu $sp, $sp, 4
	#restoring ra and fp
	addiu $sp, $sp, 4
	lw $ra, ($sp)
	addiu $sp, $sp, 4
	lw $fp, ($sp)
	#restoring 0 previously used registers
	#Moving funcRes (if exists) to next free register
	#NoneVoid function, moving its result value
	#			 __allocating reg $t0
	move	$t0,$v0
	#finished calling g
	#a Bool Func g
	bne $t0, $zero,bp_label_12
	j	bp_label_14
	#			 __freeing reg $t0
	#complicated bool exp as a function parameter
	#inserting 1 to param reg if evaluates to true
bp_label_12:
	#			 __allocating reg $t0
	li	$t0,1
	j	bp_label_13
	#inserting 0 to param reg if evaluates to false
bp_label_14:
	#			 __allocating reg $t0
	li	$t0,0
bp_label_13:
	#preparing to call f
	sw $fp, ($sp)
	addiu $sp, $sp, -4
	sw $ra, ($sp)
	addiu $sp, $sp, -4
	#pushing reg $t0 to stack
	sw $t0, ($sp)
	addiu $sp, $sp, -4
	#			 __freeing reg $t0
	move	$fp,$sp
	#jumping to func: f
	jal	label_29
	#returned from func: f
	#poping 1 params from stack 
	addiu $sp, $sp, 4
	#restoring ra and fp
	addiu $sp, $sp, 4
	lw $ra, ($sp)
	addiu $sp, $sp, 4
	lw $fp, ($sp)
	#restoring 0 previously used registers
	#Moving funcRes (if exists) to next free register
	#finished calling f
	#marker Label
bp_label_15:
	#func header store regs before call
	#stored 0 registers
	#func header store regs before call
	#stored 0 registers
	#exp derived false
	#Moving False into new Param Reg
	#			 __allocating reg $t0
	li	$t0,0
	#preparing to call g
	sw $fp, ($sp)
	addiu $sp, $sp, -4
	sw $ra, ($sp)
	addiu $sp, $sp, -4
	#pushing reg $t0 to stack
	sw $t0, ($sp)
	addiu $sp, $sp, -4
	#			 __freeing reg $t0
	move	$fp,$sp
	#jumping to func: g
	jal	label_135
	#returned from func: g
	#poping 1 params from stack 
	addiu $sp, $sp, 4
	#restoring ra and fp
	addiu $sp, $sp, 4
	lw $ra, ($sp)
	addiu $sp, $sp, 4
	lw $fp, ($sp)
	#restoring 0 previously used registers
	#Moving funcRes (if exists) to next free register
	#NoneVoid function, moving its result value
	#			 __allocating reg $t0
	move	$t0,$v0
	#finished calling g
	#a Bool Func g
	bne $t0, $zero,bp_label_16
	j	bp_label_18
	#			 __freeing reg $t0
	#complicated bool exp as a function parameter
	#inserting 1 to param reg if evaluates to true
bp_label_16:
	#			 __allocating reg $t0
	li	$t0,1
	j	bp_label_17
	#inserting 0 to param reg if evaluates to false
bp_label_18:
	#			 __allocating reg $t0
	li	$t0,0
bp_label_17:
	#preparing to call f
	sw $fp, ($sp)
	addiu $sp, $sp, -4
	sw $ra, ($sp)
	addiu $sp, $sp, -4
	#pushing reg $t0 to stack
	sw $t0, ($sp)
	addiu $sp, $sp, -4
	#			 __freeing reg $t0
	move	$fp,$sp
	#jumping to func: f
	jal	label_29
	#returned from func: f
	#poping 1 params from stack 
	addiu $sp, $sp, 4
	#restoring ra and fp
	addiu $sp, $sp, 4
	lw $ra, ($sp)
	addiu $sp, $sp, 4
	lw $fp, ($sp)
	#restoring 0 previously used registers
	#Moving funcRes (if exists) to next free register
	#finished calling f
bp_label_19:
	#Adding an extre return just in case
	move	$sp,$fp
	jr	$ra
