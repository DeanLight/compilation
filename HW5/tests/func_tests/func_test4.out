Loaded: ./exceptions.s
apple 
M boy
girl
F girl
